File: web/js/src/FFmpegKitOutputBuffer.js
/*
 * Copyright (c) 2026 Taner Sener
 *
 * This file is part of FFmpegKitNext.
 *
 * FFmpegKitNext is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * FFmpegKitNext is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General License for more details.
 *
 * You should have received a copy of the GNU Lesser General License
 * along with FFmpegKitNext. If not, see <http://www.gnu.org/licenses/>.
 */

import {getFactory} from './FFmpegKitFactory.js';

/** Seekable in-memory output; read the produced bytes with toByteArray() afterward. */
export class FFmpegKitOutputBuffer {
    constructor(handle, url) {
        this._handle = handle;
        this._url = url;
    }

    /**
     * Creates a seekable output buffer.
     *
     * Positional parameters, matching the React Native plugin's
     * FFmpegKitOutputBuffer.create().
     *
     * @param {string} extension optional output format hint, e.g. "mp4"
     * @param {?number} initialCapacity optional starting size in bytes; must be paired with maxCapacity
     * @param {?number} maxCapacity optional upper bound in bytes; must be paired with initialCapacity
     * @returns {Promise<FFmpegKitOutputBuffer>} buffer wrapper containing the generated URL
     */
    static async create(extension = '', initialCapacity = null, maxCapacity = null) {
        const {handle, url} = await getFactory().ioCreate('outputBuffer', {
            extension,
            initialCapacity,
            maxCapacity,
        });
        return new FFmpegKitOutputBuffer(handle, url);
    }

    /** @returns {string} ffkitmem: URL to use as an FFmpeg output */
    getUrl() {
        return this._url;
    }

    /** @returns {Promise<number>} number of bytes currently held by the output buffer */
    getSize() {
        return getFactory().ioGetSize(this._handle);
    }

    /** @returns {Promise<Uint8Array>} the bytes written so far. */
    toByteArray() {
        return getFactory().ioOutputBytes(this._handle);
    }

    /** @returns {Promise<void>} resolves after the worker releases the buffer */
    close() {
        return getFactory().ioClose(this._handle);
    }
}

File: web/src/FFmpegKitOutputBuffer.cpp
/*
 * Copyright (c) 2026 Taner Sener
 *
 * This file is part of FFmpegKitNext.
 *
 * FFmpegKitNext is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * FFmpegKitNext is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General License for more details.
 *
 * You should have received a copy of the GNU Lesser General License
 * along with FFmpegKitNext. If not, see <http://www.gnu.org/licenses/>.
 */

#include "FFmpegKitOutputBuffer.h"
#include "FFmpegKitConfig.h"
#include "FFmpegKitProtocolUrl.h"
#include <stdexcept>

static const long DefaultInitialCapacity = 4096;

std::shared_ptr<ffmpegkit::FFmpegKitOutputBuffer>
ffmpegkit::FFmpegKitOutputBuffer::create(const std::string &extension) {
    return ffmpegkit::FFmpegKitOutputBuffer::create(
        extension, DefaultInitialCapacity, 0);
}

std::shared_ptr<ffmpegkit::FFmpegKitOutputBuffer>
ffmpegkit::FFmpegKitOutputBuffer::create(const std::string &extension,
                                         const long initialCapacity,
                                         const long maxCapacity) {
    if (initialCapacity < 0 || maxCapacity < 0) {
        throw std::invalid_argument("capacities must not be negative");
    }

    long id = ffmpegkit::FFmpegKitConfig::registerFFmpegKitOutputBuffer(
        initialCapacity, maxCapacity);
    if (id == 0) {
        throw std::runtime_error("Failed to register FFmpegKit output buffer.");
    }

    return std::shared_ptr<ffmpegkit::FFmpegKitOutputBuffer>(
        new ffmpegkit::FFmpegKitOutputBuffer(id, extension));
}

ffmpegkit::FFmpegKitOutputBuffer::~FFmpegKitOutputBuffer() { close(); }

ffmpegkit::FFmpegKitOutputBuffer::FFmpegKitOutputBuffer(
    const long id, const std::string &extension)
    : _id{id}, _url{ffmpegkit::buildFFmpegKitUrl("ffkitmem", id, extension)},
      _closed{false} {}

std::string ffmpegkit::FFmpegKitOutputBuffer::getUrl() {
    ensureOpen();
    return _url;
}

long ffmpegkit::FFmpegKitOutputBuffer::getSize() {
    ensureOpen();
    return ffmpegkit::FFmpegKitConfig::getFFmpegKitBufferSize(_id);
}

std::shared_ptr<std::vector<uint8_t>>
ffmpegkit::FFmpegKitOutputBuffer::toByteArray() {
    ensureOpen();
    auto output = ffmpegkit::FFmpegKitConfig::getFFmpegKitOutputBuffer(_id);
    if (output == nullptr) {
        throw std::runtime_error("Failed to read FFmpegKit output buffer.");
    }

    return output;
}

void ffmpegkit::FFmpegKitOutputBuffer::close() {
    if (!_closed) {
        ffmpegkit::FFmpegKitConfig::unregisterFFmpegKitBuffer(_id);
        _closed = true;
    }
}

void ffmpegkit::FFmpegKitOutputBuffer::ensureOpen() {
    if (_closed) {
        throw std::runtime_error("FFmpegKit output buffer is closed.");
    }
}
